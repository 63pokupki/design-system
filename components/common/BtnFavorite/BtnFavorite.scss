$BtnFavoriteColorBackground: #fff;
$BtnFavoriteColorPrimary: #36a6f2;
$BtnFavoriteColorActive: #ff3868;
$BtnFavoriteWidth: 24px;
$BtnFavoriteHeight: 24px;
$BtnFavoriteFontsize: $BtnFavoriteWidth / 2;

@keyframes BtnFavoriteIsLoading {
    0% {
        opacity: 1;
    }

    50% {
        opacity: 0.1;
    }

    100% {
        opacity: 1;
    }
}

.BtnFavorite {
    @extend %reset-button;

    background-color: $BtnFavoriteColorBackground;
    width: $BtnFavoriteWidth;
    height: $BtnFavoriteHeight;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;

    &.is-loading {
        .BtnFavorite__icon {
            animation: BtnFavoriteIsLoading 1.2s ease-in-out infinite;
            color: $BtnFavoriteColorActive;
        }

        &:active {
            .BtnFavorite__icon {
                transform: none;
            }
        }
    }

    &.is-active {
        .BtnFavorite__icon {
            color: $BtnFavoriteColorActive;
        }
    }

    &:active {
        .BtnFavorite__icon {
            transform: scale(0.9);
        }
    }

    &:hover {
        .BtnFavorite__icon {
            color: $BtnFavoriteColorActive;
        }
    }

    &__icon {
        padding-top: 1px; // для выравнивания
        color: $BtnFavoriteColorPrimary;
        font-size: $BtnFavoriteFontsize;
    }
}
