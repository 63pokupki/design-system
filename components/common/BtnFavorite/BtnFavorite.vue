<script>
export default {
    name: 'BtnFavorite',
    functional: true,
    props: {
        /** Активное, нажатое состояние */
        isActive: {
            type: Boolean,
            default: false,
        },
        /** Состояние загрузки */
        isLoading: {
            type: Boolean,
            default: false,
        },
    },
    render: (h, { props, data, listeners }) => {
        const { isActive, isLoading } = props;

        const onClick = () => {
            if (!isLoading && listeners.click) {
                listeners.click();
            }
        };

        return (
            <button
                onClick={onClick}
                aria-label="Добавить в избранное"
                class={[
                    'BtnFavorite',
                    { 'is-active': isActive, 'is-loading': isLoading },
                    data.class,
                    data.staticClass,
                ]}
            >
                <i
                    class={[
                        'BtnFavorite__icon ds-icon',
                        {
                            'icon-heart-bold': !isActive,
                            'icon-heart-filled': isActive,
                        },
                    ]}
                ></i>
            </button>
        );
    },
};
</script>

<style lang="scss" scoped>
@import './BtnFavorite.scss';
</style>
