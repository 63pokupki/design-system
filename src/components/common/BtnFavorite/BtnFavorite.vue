<template>
  <button
    @click="onClick"
    class="BtnFavorite"
    :class="{'is-active':isActive, 'is-loading': isLoading}"
  >
    <i
      class="BtnFavorite__icon ds-icon"
      :class="{'icon-heart-bold': !isActive, 'icon-heart-filled': isActive}"
    ></i>
  </button>
</template>

<script>
export default {
  name: "BtnFavorite",
  props: {
    isActive: {
      type: Boolean,
      default: false,
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    onClick() {
      if (!this.isLoading) {
        this.$emit("click");
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import "./BtnFavorite.scss";
</style>