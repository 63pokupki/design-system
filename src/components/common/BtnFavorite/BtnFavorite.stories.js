import "@/components/resources";

import BtnFavorite from "./BtnFavorite.vue";
import { action } from "@storybook/addon-actions";

export default {
    title: "Общие/BtnFavorite",
    component: BtnFavorite
};

export const normal = () => ({
    components: { BtnFavorite },
    methods: {
        onClick: action("click")
    },
    template: "<BtnFavorite @click='onClick'/>"
});

export const isLoading = () => ({
    components: { BtnFavorite },
    methods: {
        onClick: action("click")
    },
    template: "<BtnFavorite @click='onClick' :is-loading='true'/>"
});

export const isActive = () => ({
    components: { BtnFavorite },
    methods: {
        onClick: action("click")
    },
    template: "<BtnFavorite @click='onClick' :is-active='true'/>"
});
