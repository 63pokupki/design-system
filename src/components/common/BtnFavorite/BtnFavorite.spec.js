import { mount, createLocalVue } from '@vue/test-utils';
import BtnFavorite from './BtnFavorite.vue';

/** Фабрика создания компонента */
const mountComponent = (context = {}) => {
    const localVue = createLocalVue();
    return mount(BtnFavorite, {
        ...context,
        localVue,
    });
};

describe('BtnFavorite.vue', () => {
    it('Инициализация', () => {
        const wrapper = mountComponent();

        expect(wrapper.exists()).toBeTruthy();
    });

    it('Снимок без состояния', () => {
        const wrapper = mountComponent();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка изменения входных параметров и вида компонента', () => {
        const wrapper = mountComponent({
            propsData: { isActive: true, isLoading: true },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Обработка нажатия', () => {
        const fnMock = jest.fn();

        const wrapper = mountComponent({
            listeners: {
                click: fnMock,
            },
        });

        wrapper.trigger('click');

        expect(fnMock).toBeCalled();
    });
});
