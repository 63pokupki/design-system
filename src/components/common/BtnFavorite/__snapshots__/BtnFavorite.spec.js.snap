// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BtnFavorite.vue Проверка изменения входных параметров и вида компонента 1`] = `<button aria-label="Добавить в избранное" class="BtnFavorite is-active is-loading"><i class="BtnFavorite__icon ds-icon icon-heart-filled"></i></button>`;

exports[`BtnFavorite.vue Снимок без состояния 1`] = `<button aria-label="Добавить в избранное" class="BtnFavorite"><i class="BtnFavorite__icon ds-icon icon-heart-bold"></i></button>`;
